contrib/sarplus/README.md

```diff
@@ -0,0 +1,191 @@
+* shared read-only memory mapping allows us to re-use the same memory from multiple python executors on the same worker node
+* partition the input test users and past seen items by users, allowing for scale out
+* perform as much of the work as possible in PySpark (way simpler)
+* top-k computation
+** reverse the join by summing reverse joining the users past seen items with any related items
+** make sure to always just keep top-k items in-memory
+** use standard join using binary search between users past seen items and the related items
+
+![Image of sarplus top-k recommendation optimization](images/sarplus_udf.svg) 
+
+# Usage
+
+```python
+import pandas as pd
+from pysarplus import SARPlus
+
+# spark dataframe with user/item/rating/optional timestamp tuples
+train_df = spark.createDataFrame(
+      pd.DataFrame({
+        'user_id': [1, 1, 2, 3, 3],
+        'item_id': [1, 2, 1, 1, 3],
+        'rating':  [1, 1, 1, 1, 1],
+    }))
+   
+# spark dataframe with user/item tuples
+test_df = spark.createDataFrame(
+      pd.DataFrame({
+        'user_id': [1, 3],
+        'item_id': [1, 3],
+        'rating':  [1, 1],
+    }))
+    
+model = SARPlus(spark, col_user='user_id', col_item='item_id', col_rating='rating', col_timestamp='timestamp')
+model.fit(train_df, similarity_type='jaccard')
+
+
+model.recommend_k_items(test_df, 'sarplus_cache', top_k=3).show()
+
+# For databricks
+# model.recommend_k_items(test_df, 'dbfs:/mnt/sarpluscache', top_k=3).show()
+```
+
+## Jupyter Notebook
+
+Insert this cell prior to the code above.
+
+```python
+import os
+
+SUBMIT_ARGS = "--packages microsoft:sarplus:0.2.6 pyspark-shell"
+os.environ["PYSPARK_SUBMIT_ARGS"] = SUBMIT_ARGS
+
+from pyspark.sql import SparkSession
+
+spark = (
+    SparkSession.builder.appName("sample")
+    .master("local[*]")
+    .config("memory", "1G")
+    .config("spark.sql.shuffle.partitions", "1")
+    .config("spark.sql.crossJoin.enabled", True)
+    .config("spark.ui.enabled", False)
+    .getOrCreate()
+)
+```
+
+## PySpark Shell
+
+```bash
+pip install pysarplus
+pyspark --packages microsoft:sarplus:0.2.6 --conf spark.sql.crossJoin.enabled=true
+```
+
+## Databricks
+
+One must set the crossJoin property to enable calculation of the similarity matrix (Clusters / &lt; Cluster &gt; / Configuration / Spark Config)
+
+```
+spark.sql.crossJoin.enabled true
+```
+
+1. Navigate to your workspace 
+2. Create library
+3. Under 'Source' select 'Maven Coordinate'
+4. Enter 'microsoft:sarplus:0.2.5' or 'microsoft:sarplus:0.2.6' if you're on Spark 2.4.1
+5. Hit 'Create Library'
+6. Attach to your cluster
+7. Create 2nd library
+8. Under 'Source' select 'Upload Python Egg or PyPI'
+9. Enter 'pysarplus'
+10. Hit 'Create Library'
+
+This will install C++, Python and Scala code on your cluster.
+
+You'll also have to mount shared storage

```

Comment:
This may be optional - we can always use a local storage on ADB for caching the results, thought I am not sure if there is any downside.

Reply:
I don't think local storage will work as each node memory maps the distributed file..

Reply:
By local I mean `dbfs`

Reply:
if you refer to /dbfs/mnt/foo it has the big issue that it has issues with files larger than 2GB (I think it's backed by fuse